File: PE-common/src/main/java/nl/adaptivity/process/processModel/engine/XmlCondition.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.processModel.engine;

import net.devrieze.util.StringUtil;
import nl.adaptivity.process.ProcessConsts.Engine;
import nl.adaptivity.process.processModel.Condition;
import nl.adaptivity.xml.*;
import org.jetbrains.annotations.NotNull;

import javax.xml.namespace.QName;


/**
 * Class encapsulating a condition.
 *
 * @author Paul de Vrieze
 */
public class XmlCondition implements XmlSerializable, Condition {

  private final String mCondition;

  public XmlCondition(final String condition) {
    mCondition = condition;
  }

  @Override
  public void serialize(@NotNull final XmlWriter out) throws XmlException {
    XmlWriterUtil.writeSimpleElement(out, new QName(Engine.NAMESPACE, ELEMENTLOCALNAME, Engine.NSPREFIX), getCondition());
  }

  @NotNull
  public static XmlCondition deserialize(@NotNull final XmlReader reader) throws XmlException {
    final CharSequence condition = XmlReaderUtil.readSimpleElement(reader);
    return new XmlCondition(StringUtil.toString(condition));
  }

  /* (non-Javadoc)
   * @see nl.adaptivity.process.processModel.engine.Condition#getCondition()
   */
  @Override
  public String getCondition() {
    return mCondition;
  }

}

File: PE-common/src/commonMain/kotlin/nl/adaptivity/process/processModel/engine/XmlCondition.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.processModel.engine

import kotlinx.serialization.Serializable
import kotlinx.serialization.Serializer
import kotlinx.serialization.*
import kotlinx.serialization.builtins.nullable
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import nl.adaptivity.process.ProcessConsts
import nl.adaptivity.process.ProcessConsts.Engine
import nl.adaptivity.process.processModel.Activity
import nl.adaptivity.process.processModel.Condition
import nl.adaptivity.serialutil.DelegatingSerializer
import nl.adaptivity.serialutil.withName
import nl.adaptivity.xmlutil.*
import nl.adaptivity.xmlutil.serialization.XmlSerialName


/**
 * Class encapsulating a condition.
 *
 * @author Paul de Vrieze
 */
@OptIn(ExperimentalSerializationApi::class)
@Serializable(/*XmlCondition.Companion::class*/)
@XmlSerialName(Condition.ELEMENTLOCALNAME, Engine.NAMESPACE, Engine.NSPREFIX)
class XmlCondition(override val condition: String, override val label: String? = null) : Condition {

    override fun toString(): String {
        return "XmlCondition(condition='$condition')"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is XmlCondition) return false

        if (condition != other.condition) return false

        return true
    }

    override fun hashCode(): Int {
        return condition.hashCode()
    }

    companion object : DelegatingSerializer<XmlCondition, String>(String.serializer()) {

        override val descriptor: SerialDescriptor =
            PrimitiveSerialDescriptor("condition", PrimitiveKind.STRING)

        override fun fromDelegate(delegate: String): XmlCondition {
            return XmlCondition(delegate)
        }

        override fun XmlCondition.toDelegate(): String {
            return condition
        }
    }
}
